drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,225 @@
+		 * being overwritten by another threshold configuration.
+		 */
+		if (closest_mv == code_cfg->press_mv) {
+			key_data->curr_state = 1;
+		}
+	}
+
+	/*
+	 * Report the key event if the key state has changed.
+	 */
+
+	for (uint32_t i = 0; i < cfg->key_cnt; i++) {
+		key_data = &cfg->key_data[i];
+
+		if (key_data->last_state != key_data->curr_state) {
+			LOG_DBG("Report event %s %d, code=%d", dev->name, key_data->curr_state,
+				key_data->zephyr_code);
+			input_report_key(dev, key_data->zephyr_code, key_data->curr_state, true,
+					 K_FOREVER);
+			key_data->last_state = key_data->curr_state;
+		}
+
+		/*
+		 * Reset the state so that it can be updated in the next
+		 * iteration.
+		 */
+		key_data->curr_state = 0;
+	}
+}
+
+static void adc_keys_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+	struct adc_keys_data *data = CONTAINER_OF(dwork, struct adc_keys_data, dwork);
+	const struct device *dev = data->self;
+	const struct adc_keys_config *cfg = dev->config;
+
+	adc_keys_process(dev);
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+}
+
+static int adc_keys_init(const struct device *dev)
+{
+	const struct adc_keys_config *cfg = dev->config;
+	struct adc_keys_data *data = dev->data;
+	int ret;
+
+	if (!device_is_ready(cfg->channel.dev)) {
+		LOG_ERR("ADC controller device %s not ready", cfg->channel.dev->name);
+		return -ENODEV;
+	}
+
+	ret = adc_channel_setup_dt(&cfg->channel);
+	if (ret) {
+		LOG_ERR("ADC channel setup failed %d", ret);
+		return ret;
+	}
+
+	adc_sequence_init_dt(&cfg->channel, &data->seq);
+
+	data->self = dev;
+	k_work_init_delayable(&data->dwork, adc_keys_work_handler);
+
+#if defined(CONFIG_INPUT_LOG_LEVEL_DBG)
+	for (uint32_t i = 0; i < cfg->code_cnt; i++) {
+		LOG_DBG("* code %d: key_index=%d threshold=%d mV code=%d", i,
+			cfg->code_cfg[i].key_index, cfg->code_cfg[i].press_mv,
+			cfg->key_data[cfg->code_cfg[i].key_index].zephyr_code);
+	}
+#endif /* CONFIG_INPUT_LOG_LEVEL_DBG */
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+
+	return 0;
+}
+
+#define ADC_KEYS_CODE_CFG_CODE(node_id, prop, idx)                                                 \
+	{                                                                                          \
+		.key_index = DT_NODE_CHILD_IDX(node_id) /* include disabled nodes */,              \
+		.press_mv = DT_PROP_BY_IDX(node_id, prop, idx),                                    \
+	}
+
+#define ADC_KEYS_CODE_CFG(node_id)                                                                 \
+	DT_FOREACH_PROP_ELEM_SEP(node_id, press_thresholds_mv, ADC_KEYS_CODE_CFG_CODE, (,))
+
+#define ADC_KEYS_KEY_DATA(node_id)                                                                 \
+	{                                                                                          \
+		.zephyr_code = DT_PROP(node_id, zephyr_code), .last_state = 0, .curr_state = 0,    \
+	}
```

Comment:
how about moving zephyr_code in its own array? I think you could have a `const uint16_t *codes` for the codes, that is going to be `key_cnt` sized, then you only have to initialize that statically and you can leave `key_data` being zero initialized automatically. That will save another bit of flash and another bit of ram. :-)